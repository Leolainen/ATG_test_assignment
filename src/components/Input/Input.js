import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import styles from "./styles.module.css";

const Input = React.forwardRef(function Input(props, ref) {
  const { children, className, label, cta, ...other } = props;

  return (
    <div className={classnames(styles.root, className)}>
      {label && <label className={styles.label}>{label}</label>}

      <div className={styles.inner}>
        <input ref={ref} className={styles.input} {...other} />

        {cta && <div className={styles.cta}>{cta}</div>}
      </div>
    </div>
  );
});

Input.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string,
  cta: PropTypes.element,
};

Input.uiName = "Input";

export default Input;
