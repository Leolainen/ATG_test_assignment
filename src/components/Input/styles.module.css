.root {
}

.inner {
  position: relative;
  display: flex;
  flex-wrap: nowrap;
  vertical-align: middle;
}

.input {
  width: 100%;
  cursor: text;
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;
  border: 1px solid var(--border);
  padding: 6px;
  border-radius: 3px;
  box-shadow: 0 0.7px 1.6px rgba(0, 0, 0, 0.031),
    0 2px 4.4px rgba(0, 0, 0, 0.045), 0 4.8px 10.6px rgba(0, 0, 0, 0.059),
    0 16px 35px rgba(0, 0, 0, 0.09);
}

.input:focus {
  outline-color: var(--harmony-2);
}

.label {
  display: block;
  margin-bottom: 4px;
}

.cta {
  position: absolute;
  /**
    Some magic pixels so search looks fancy
  */
  right: 2px;
  top: 1px;
}
