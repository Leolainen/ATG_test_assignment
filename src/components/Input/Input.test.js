import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import Input from "./Input";

describe("<Input />", () => {
  afterEach(cleanup);

  it("should render label text", () => {
    const { getByText } = render(<Input label="label" />);
    const label = getByText("label");

    expect(label).toBeInTheDocument();
  });

  describe("CTA", () => {
    it("should render CTA", () => {
      const { getByText } = render(<Input cta={<button>click me</button>} />);
      const button = getByText("click me");

      expect(button).toBeInTheDocument();
    });

    it("should fire onClick callback when clicked", () => {
      const mockFn = jest.fn();
      const { getByText } = render(
        <Input cta={<button onClick={mockFn}>click me</button>} />
      );
      const button = getByText("click me");
      fireEvent.click(button);

      expect(mockFn).toHaveBeenCalled();
    });
  });

  describe("Snapshots", () => {
    it("should render default", () => {
      const { asFragment } = render(<Input />);

      expect(asFragment()).toMatchSnapshot();
    });

    it("should render with label and cta", () => {
      const { asFragment } = render(
        <Input label="label" cta={<button>click me</button>} />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
