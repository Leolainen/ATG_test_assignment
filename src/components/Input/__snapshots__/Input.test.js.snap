// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> Snapshots should render default 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="inner"
    >
      <input
        class="input"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Input /> Snapshots should render with label and cta 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <label
      class="label"
    >
      label
    </label>
    <div
      class="inner"
    >
      <input
        class="input"
      />
      <div
        class="cta"
      >
        <button>
          click me
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
